lib/collection/src/collection_manager/optimizers/config_mismatch_optimizer.rs

```diff
@@ -0,0 +1,488 @@
+};
+use segment::types::{HnswConfig, Indexes, QuantizationConfig, SegmentType, VECTOR_ELEMENT_SIZE};
+
+use crate::collection_manager::holders::segment_holder::{LockedSegmentHolder, SegmentId};
+use crate::collection_manager::optimizers::segment_optimizer::{
+    OptimizerThresholds, SegmentOptimizer,
+};
+use crate::config::CollectionParams;
+use crate::operations::config_diff::DiffConfig;
+
+/// Looks for segments having a mismatch between configured and actual parameters
+///
+/// For example, a user may change the HNSW parameters for a collection. A segment that was already
+/// indexed with different parameters now has a mismatch. This segment should be optimized (and
+/// indexed) again in order to update the effective configuration.
+pub struct ConfigMismatchOptimizer {
+    thresholds_config: OptimizerThresholds,
+    segments_path: PathBuf,
+    collection_temp_dir: PathBuf,
+    collection_params: CollectionParams,
+    hnsw_config: HnswConfig,
+    quantization_config: Option<QuantizationConfig>,
+    telemetry_durations_aggregator: Arc<Mutex<OperationDurationsAggregator>>,
+}
+
+impl ConfigMismatchOptimizer {
+    pub fn new(
+        thresholds_config: OptimizerThresholds,
+        segments_path: PathBuf,
+        collection_temp_dir: PathBuf,
+        collection_params: CollectionParams,
+        hnsw_config: HnswConfig,
+        quantization_config: Option<QuantizationConfig>,
+    ) -> Self {
+        ConfigMismatchOptimizer {
+            thresholds_config,
+            segments_path,
+            collection_temp_dir,
+            collection_params,
+            hnsw_config,
+            quantization_config,
+            telemetry_durations_aggregator: OperationDurationsAggregator::new(),
+        }
+    }
+
+    /// Calculates and HNSW config that should be used for a given vector
+    /// with current configuration.
+    ///
+    /// Takes vector-specific HNSW config (if any) and merges it with the collection-wide config.
+    fn get_required_hnsw_config(&self, vector_name: &str) -> Cow<HnswConfig> {
+        let target_hnsw_collection = &self.hnsw_config;
+        // Select vector specific target HNSW config
+        let target_hnsw_vector = self
+            .collection_params
+            .vectors
+            .get_params(vector_name)
+            .and_then(|vector_params| vector_params.hnsw_config)
+            .map(|vector_hnsw| vector_hnsw.update(target_hnsw_collection))
+            .and_then(|hnsw| match hnsw {
+                Ok(hnsw) => Some(hnsw),
+                Err(err) => {
+                    log::warn!(
+                        "Failed to merge collection and vector HNSW config, ignoring: {err}"
+                    );
+                    None
+                }
+            });
+        match target_hnsw_vector {
+            Some(target_hnsw) => Cow::Owned(target_hnsw),
+            None => Cow::Borrowed(target_hnsw_collection),
+        }
+    }
+
+    fn worst_segment(
```

Comment:
Maybe a doc comment will make it easier to skim through this function.

Also, if we are returning `Vec<_>` wouldn't it make sense to use plural?
```suggestion
    fn worst_segments(
```

Reply:
Good point, this is a bit confusing.

Usually, just a single segment is returned. However, there are some edge cases where multiple are returned when it is estimated combining them is better.

This function name originates from the other optimizers. We might have to do another PR to improve naming for all of them.